src/tcp_tunnel_tools.js

```diff
@@ -81,26 +69,19 @@ function createTunnel(proxyUrl, targetHost, providedOptions = {}, callback) {
 
 module.exports.createTunnel = createTunnel;
 
-function closeTunnel(serverPath, closeConnections, callback) {
-    const [hostname, port] = serverPath.split(':');
-    if (!hostname) throw new Error('serverPath must contain hostname');
-    if (!port) throw new Error('serverPath must contain port');
```

Comment:
I'm not saying this validation is great, but `closeTunnel` is public API. So we should validate this somehow or document the shape of `address` properly and enforce its type.

Reply:
> So we should validate this somehow

This return `false` if the server does not exist. If anybody wants to reassure an actual server was closed they should make an assertion.

Reply:
It's not about making sure that a server was closed, but knowing that you passed in the correct argument.

Reply:
> It's not about making sure that a server was closed, but knowing that you passed in the correct argument.

I don't understand how relevant this is. The purpose of this function is to close a server.
Can you explain the reasoning please?

Reply:
When I pass in an invalid address, it tells me that the server is closed, instead of telling me that the address is invalid. The previous validation made sure that you're at least passing in something reasonable.

Reply:
> When I pass in an invalid address, **it tells me that the server is closed**, instead of telling me that the address is invalid.

With the previous implementation on `master`, it returns `false` as well if you provide a valid address but [the server may haven't existed at all](https://github.com/apify/proxy-chain/blob/0043ac9db2c04fffdd53fe7ee7d3a11183648555/src/tcp_tunnel_tools.js#L94). Following your thinking, it should throw an error because the server can't be closed. And to check if the server was already closed, we would have to store it, which would do a memory leak. So the solution would be to expose it, so we don't have to care about storing it.

You get it wrong. `false` doesn't tell you that the server is closed, it says that it failed to do so.

As I've [already mentioned](https://github.com/apify/proxy-chain/pull/158#discussion_r710264562), an `error` is passed to the callback only and only if the server was already closed, which is never the case here.

> The previous validation made sure that you're at least passing in something reasonable.

How does this explain that it should throw? We want to close a server, not to check if the server actually exists. If the user code assumes the server exists, they should assert that this returns `true` - and this already validates the argument.

---

I propose to return the actual server and get rid of `closeTunnel`.